Adds table name to the field proto

bbeaudreault opened the pull request:
https://dev.mysql.com/doc/refman/5.7/en/c-api-data-structures.html 

The `table` and `table_length` fields are available in the C api but are not passed through to clients for use. One of our core java client libraries depends on http://docs.oracle.com/javase/7/docs/api/java/sql/ResultSetMetaData.html#getTableName(int), which in the existing Vitess Driver was hardcoded to return null.

This PR adds `table` to the `Field` proto, populates it in the `mysql.Fields()` function, and allows access in the Java client.  All tests updated to account for the new field, and succeeding.

I've also tested this in our deployment.  However, I've not been able to get the full `make test` to work in my local dev yet, this was tested with `make site_test`.

This is my first foray into both go and vitess, so feedback welcome.

alainjobart commented:
We wanted to make this extra field optional, to not add too much network overhead for the apps that don't need it. So we want to add a boolean to 'ExecuteOptions' to include that info or not. ExecuteOptions already has exclude_field_names to not add field names, it could also have an 'include_table_names' to include the fields. What do you think?

bbeaudreault commented:
Alright, I'll add that

sougou commented:
FTR, Jiten is working on supporting this (and other necessary extras) as part of the MySQL protocol work.

bbeaudreault commented:
Have been away for a few days but am getting back to this. I spoke to @jvaidya and we agreed that I can take over adding the extra fields in this PR instead of his, as it was orthogonal to his work.

I'm wrapping up the inclusion of an exclude_field_metadata in Query.ExecuteOptions, following the convention of exclude_field_names. The goal would be for my new option to supercede the now-deprecated old one, as I don't think it makes sense to have a separate `exclude_` field for every field metadata type.

Once I've got that working and tested in my environment, I'll add the other mysql protocol fields to the response as well, similarly excluded when `exclude_field_metadata` is set.

Aside: There was a problem with our corporate CLA and we're hoping to have that resolved next week.

bbeaudreault commented:
I think this is ready for a review. Here's a couple notes:

- This PR only adds the new fields to the response, and the exclude_field_metadata option
- I will add a separate PR for JDBC support of these fields (code is almost ready, adding tests)
- All tests have been updated to account for the new fields. Tests were added to ensure exclude_field_metadata and exclude_field_names have basically the same effect (for backwards compatibility)
- Flags were added to the proto to support additional features in the JDBC, however...
- Flags add a bit of complexity to the tests -- in local testing (which was using MySQL5.7), results would include an additional `NO_DEFAULT_VALUE_FLAG`. But in travis CI (which uses mariadb), it did not include that flag. This caused mismatches in the endtoend tests. This is why I added explicit default values to all fields in main_test.go. I'm open to other suggestions here.

alainjobart commented:
Sorry for the late review, I wanted to check with the team on this, and it took me forever. :(

The main comment we have is we think we should keep the query options as follow:
- exclude_field_names bool: do not deprecate it, keep it.
- change your exclude_field_metadata into an include_field_metadata.
- Field then has:
+ Name string
+ Type
+ a bunch of other fields.

By default, we return both Name and Type. To be efficient, and if a client wants to at high QPS, they can remove the Name (but if nobody uses that, we'll just remove that flag). To be complete, and if a client wants to, they can add the extra metadata. The Java client may always do that.

No extra query option does what we do now, which we think most users want.

What do you think?

bbeaudreault commented:
@alainjobart -- this sounds good to me, however i feel like it's weird to have both an exclude and an include like field. It would probably be clearer to use exclude for both, but maybe default `exclude_field_metadata` to true.  Thoughts?

alainjobart commented:
With proto3 you can't have a boolean default to true... it is false by default.

I agree having both exclude and include flags is not ideal. But they somewhat match the fact that what most people want is one included and the other one excluded...

Let me check if anyone is using the exclude_field_name one, and if not we could just remove that one.

bbeaudreault commented:
@alainjobart I pushed a commit which changes to an Enum. It compiles, but the tests will *not* succeed currently.

Let me know what you think of this, and I can either revert or get the tests working and clean it up

bbeaudreault commented:
FYI @jvaidya you'll want to rebase your mysql proxy on top of this once it gets merged

jvaidya commented:
Yep, I have been following this thread with interest, and will do so when
your code lands. Thanks.



On Thu, Jan 5, 2017 at 11:53 AM, Bryan Beaudreault <notifications@github.com
> wrote:

> FYI @jvaidya <https://github.com/jvaidya> you'll want to rebase your
> mysql proxy on top of this once it gets merged
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/youtube/vitess/pull/2381#issuecomment-270740413>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEXuGdJsnU-x52gNah6HqCCWtUl3N7OFks5rPUoogaJpZM4LPpRk>
> .
>

alainjobart commented:
@bbeaudreault this looks great, both Sugu and I like it!

I don't think anyone is using the exclude_field_name flag. I'd be OK with you just deprecating it (by reserving the filed id in the proto and commenting it, like:

// This uses to be exclude_field_name, deprecated and replaced by the enum below...
reserved 1;

Then all you need to do is pass the enum value instead of the exclude_field_name in the APIs, and you're good. That should make the change simpler too.

Does that work for you?

bbeaudreault commented:
Works for me, thanks @alainjobart! I'll go ahead and reserve that id, then remove up the compatibility code and fix/cleanup tests. Will ping again today or tomorrow for a re-review

bbeaudreault commented:
@alainjobart this is all set. I still kept a function to handle converting a `nil` ExecuteOptions to the default value. I could move that in to StripMetadata directly if that would be cleaner.

I added an endtoend test to ensure that we properly handle that fallback case when various values of ExecuteOptions are passed into Execute/StreamExecute.  And I modified the result_test.go to handle the various possible ExecuteOption values.

I also tried updating the docs. I updated VitessApi.md, but could not figure out how to get `make build_web` to work

alainjobart commented:
Just a comment on the web site, you just need to run:
./publish-site.sh --docker
it will use the right docker images to build the site, no need to download any software.

alainjobart commented on proto/query.proto:
```diff
@@ -203,6 +239,14 @@ message Field {
 
   // vitess-defined type. Conversion function is in sqltypes package.
   Type type = 2;
+  string table = 3;
```
I would add a comment before this group of fields saying they are added only with the query option.

bbeaudreault commented:
@alainjobart done.

Thanks for the review, and also for the note about publishing the site.  I looked into doing that just now, and it a ton of site files. To avoid merge conflicts (i noticed you recently published the site too), I excluded the publish commit from this PR.  Once it's merged, someone (or I) can do a separate PR for the site

alainjobart commented:
Sounds good, I'll do the publish later.

I'll merge this as soon as travis passes.

michael-berlin commented:
@bbeaudreault Did you get pre-commit warnings when you created commits for the Go code?

For example, go lint should have complained about this in `sqltypes/result.go`:

`
// Normalizes the passed Execution Options, returning our default value if options happens to be nil
func IncludeFieldsOrDefault(options *querypb.ExecuteOptions) querypb.ExecuteOptions_IncludedFields {
`

The lint error is that the comment must start with `// IncludeFieldsOrDefault ...`.

I'm asking because I want to understand if external contributors explicitly ignore the warnings or there's a general problem with the way we install the pre-commit check in bootstrap.sh.

Ideally, nobody should ignore any precommit warning and always address them :)

Regarding the lint errors itself. I'll create a PR shortly to fix them.

(I'm bringing this up now because our internal tooling runs go lint on the changed code when we import it from GitHub.)

bbeaudreault commented:
Sorry for that @michael-berlin  -- I have not gotten any precommit warnings. I didn't know they existed, in fact.

I just checked my git config, and I'm using a global git `hooksPath` config to point at my company's git hooks.  I can try to symlink the vitess githooks into it

michael-berlin commented:
No worries :)

Did you run `bootstrap.sh`?

Do you know if the global `hooksPath` will result into ignoring the per-Git hooks?

We're enabling the hooks for the Vitess Git repo here by copying it to `.git/hooks` (relative to the local Vitess Git repository): https://github.com/youtube/vitess/blob/master/bootstrap.sh#L246

bbeaudreault commented:
@michael-berlin yes, I did run bootstrap.sh, and it did symlink the pre-commit into `.git/hooks`. The problem is I guess if you have global commit hooks (not per-repo), those take precedence.  I'm still looking into it, but I at least got the pre-commit hooks running now by symlinking them to make them get picked up by the global hooks we have configured outside.

Also, would it make sense to add running of the hooks to the travis build? That also would ensure that no one accidentally misses the hooks

michael-berlin commented:
> Also, would it make sense to add running of the hooks to the travis build? That also would ensure that no one accidentally misses the hooks

Yeah, I agree ;(

I already looked into it once and thought a custom solution wouldn't be ideal. Instead, I was considering more generic solutions which you can add as service to GitHub e.g. https://codeclimate.com/

Maybe I should re-consider to run our hooks. There are only two problems:
- I need to find out against what we should diff e.g. should we calculate the diff of a PR against master?
- sometimes there are false positives and we don't want to block the tests on that :)

bbeaudreault commented:
Running

```bash
git config core.hooksPath .git/hooks
```

Within the cloned vitess git repository adds the proper section to the `.git/config` to ensure that the hooks are run. I've verified that it overrides my company's global git hookPath, which makes sense those commit hooks don't make sense in vitess. 

Maybe we can add that call to the bootstrap.sh? Or is that too heavy handed?

michael-berlin commented:
I don't see a downside: After setting it on my workstation, the hooks still work for me.

Do you want to send a PR for the change? (Otherwise, I can also do it.)

bbeaudreault commented:
Submitted https://github.com/youtube/vitess/pull/2452
